libcudacxx/include/cuda/__floating_point/config.h

```diff
@@ -0,0 +1,526 @@
+  {
+    return _Tp{__fp_from_storage, typename _Tp::__storage_type{0xff}};
+  }
+  _CCCL_TEMPLATE(class _Tp)
+  _CCCL_REQUIRES(__is_cuda_extended_floating_point_v<_Tp>)
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr _Tp __nan()
+  {
+    return _Tp{__fp_from_storage, static_cast<typename _Tp::__storage_type>(0xff)};
+  }
+  _CCCL_TEMPLATE(class _Tp)
+  _CCCL_REQUIRES(__is_cuda_extended_floating_point_v<_Tp>)
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr bool __is_nan(const _Tp& __val)
+  {
+    return __val.__storage == static_cast<typename _Tp::__storage_type>(0xff);
+  }
+};
+
+template <class _Tp>
+_LIBCUDACXX_HIDE_FROM_ABI constexpr auto __fp_make_config_from()
+{
+  using _Up = _CUDA_VSTD::remove_cv_t<_Tp>;
+
+  if constexpr (__is_cuda_extended_floating_point_v<_Up>)
+  {
+    return typename _Up::__config_type{};
+  }
+  else if constexpr (_CUDA_VSTD::is_same_v<_Up, float>)
+  {
+    return __fp32_config{};
+  }
+  else if constexpr (_CUDA_VSTD::is_same_v<_Up, double>)
+  {
+    return __fp64_config{};
+  }
+#  if !defined(_LIBCUDACXX_HAS_NO_LONG_DOUBLE)
+  else if constexpr (_CUDA_VSTD::is_same_v<_Up, long double>)
+  {
+    return __fp_long_double_config{};
+  }
+#  endif // !_LIBCUDACXX_HAS_NO_LONG_DOUBLE
+  else if constexpr (_CUDA_VSTD::is_same_v<_Up, typename __fp16_config::__host_native_type>
+                     || _CUDA_VSTD::is_same_v<_Up, typename __fp16_config::__device_native_type>)
+  {
+    return __fp16_config{};
+  }
+  else if constexpr (_CUDA_VSTD::is_same_v<_Up, typename __fp128_config::__host_native_type>
+                     || _CUDA_VSTD::is_same_v<_Up, typename __fp128_config::__device_native_type>)
+  {
+    return __fp128_config{};
+  }
+  else if constexpr (_CUDA_VSTD::is_same_v<_Up, typename __bf16_config::__host_native_type>
+                     || _CUDA_VSTD::is_same_v<_Up, typename __bf16_config::__device_native_type>)
+  {
+    return __bf16_config{};
+  }
+#  if __STDCPP_FLOAT16_T__ == 1
+  else if constexpr (_CUDA_VSTD::is_same_v<_Up, ::std::float16_t>)
+  {
+    return __fp16_config{};
+  }
+#  endif // __STDCPP_FLOAT16_T__
+#  if __STDCPP_FLOAT32_T__ == 1
+  else if constexpr (_CUDA_VSTD::is_same_v<_Up, ::std::float32_t>)
+  {
+    return __fp32_config{};
+  }
+#  endif // __STDCPP_FLOAT32_T__
+#  if __STDCPP_FLOAT64_T__ == 1
+  else if constexpr (_CUDA_VSTD::is_same_v<_Up, ::std::float64_t>)
+  {
+    return __fp64_config{};
+  }
+#  endif // __STDCPP_FLOAT64_T__
+#  if __STDCPP_FLOAT128_T__ == 1
+  else if constexpr (_CUDA_VSTD::is_same_v<_Up, ::std::float128_t>)
+  {
+    return __fp128_config{};
+  }
+#  endif // __STDCPP_FLOAT128_T__
+#  if __STDCPP_BFLOAT16_T__ == 1
+  else if constexpr (_CUDA_VSTD::is_same_v<_Up, ::std::bfloat16_t>)
+  {
+    return __bf16_config{};
+  }
+#  endif // __STDCPP_BFLOAT16_T__
+  else
+  {
+    return __fp_invalid_config{};
+  }
+}
+
+template <class _Tp>
+using __fp_make_config_from_t = decltype(::cuda::__fp_make_config_from<_Tp>());
```

Comment:
...and when all the above config types are templates, this could get turned into a simple computation of "how many bits of X do you have"; see https://github.com/griwes/cccl/commit/c81edb4fba39e72f3eddfba23b6fd48d6c1d6151#diff-0c3dd6fef1b5e163c5549aa21479e2698be89899baffd111199bdc5bd6633645 for a general reference for how I envisioned this idea working.